add simple_pipeline (#8)
membraneframework/membrane_tutorials: get_started/simple_pipeline.md

```
##### Mac
```bash
$ brew install clang-format portaudio ffmpeg libmad pkg-config
```

### Creating a Project

While installing Elixir you'll get a bunch of useful tools. One of them is [Mix](https://hexdocs.pm/mix/Mix.html). As you can read in its documentation preface:
> Mix is a build tool that provides tasks for creating, compiling, and testing Elixir projects, managing its dependencies, and more.

Lets use it to create a project for our first Membrane adventure:

```bash
$ mix new hello --module Hello
```

Mix generator will create some files for us. Lets take a closer look into two of them:

**mix.exs** - It's an entry file for our mix project, a place where we can configure it, and set-up our dependencies. We'll do it by adding them into `deps` function:

```elixir
defp deps do
  [
	{:membrane_core, "~> 0.7.0"},
	{:membrane_file_plugin, "~> 0.6.0"},
	{:membrane_portaudio_plugin, "~> 0.7.0"},
	{:membrane_ffmpeg_swresample_plugin, "~> 0.7.1"},
	{:membrane_mp3_mad_plugin, "~> 0.7.0"}
  ]
end
```
Next we can fetch them by running
```bash
$ mix deps.get
```

**lib/hello.ex** - Here we'll be creating our pipeline.

### Our first Pipeline

> 

Reviewer: I would add a quick intro sentence stating that pipelines are one of the standard elements used by Membrane and that they function to take in some data, push it through a series of elements that transform the data and output it somewhere.

#### Pipeline behaviour

Let's start with declaring that we'll be using `Membrane.Pipeline` behaviour:

```elixir
defmodule Hello do

  use Membrane.Pipeline
  
end
```

> Using a behaviour means we are treating our module as a Membrane pipeline, so we've access to functions defined in `Membrane.Pipeline` module, and we can implement some of it's callbacks.

Reviewer: ```suggestion
Using this behaviour means we are treating our module as a Membrane Pipeline, so we will have access to functions defined in the `Membrane.Pipeline` module, and we can implement some of it's callbacks.
```

> Let's implement a first one, `handle_init/1`. As you can see in [the documentation](https://hexdocs.pm/membrane_core/Membrane.Pipeline.html#c:handle_init/1) `handle_init` takes one argument wich can be any type - it's a way to pass some arguments needed for pipeline to start. As our app one and only purpose id to play an mp3 file we can assume that the only value we need to pass into pipeline is a path for a file we want to play:

Reviewer: ```suggestion
Let's implement the first callback: `handle_init/1`. As you can see in [the documentation](https://hexdocs.pm/membrane_core/Membrane.Pipeline.html#c:handle_init/1) `handle_init` takes one argument wich can be of any type - it's a way to pass some arguments needed for the pipeline to start. As our app's one and only purpose is to play an mp3 file we can assume that the only value we need to pass into the pipeline is a path for a file we want to play:
```

```elixir
defmodule Hello do

  use Membrane.Pipeline

  @impl true
  def handle_init(path_to_mp3) do
  end
  
end
```

> The main purpose of `handle_init` callback is to prepare our pipeline. Preparing among others means that we need to specify all its elements as children and set-up linkings between them. All other options that can be specified if needed you can find in `Membrane.ParentSpec` [docs](https://hexdocs.pm/membrane_core/Membrane.ParentSpec.html).

Reviewer: ```suggestion
The main purpose of the `handle_init` callback is to prepare our pipeline. Preparing means that we need to specify all its elements as children and set up links between those children to describe the order in which data will flow through the pipeline. You can find all other options that can be specified if needed in the `Membrane.ParentSpec` [docs](https://hexdocs.pm/membrane_core/Membrane.ParentSpec.html).
```

> The main purpose of `handle_init` callback is to prepare our pipeline. Preparing among others means that we need to specify all its elements as children and set-up linkings between them. All other options that can be specified if needed you can find in `Membrane.ParentSpec` [docs](https://hexdocs.pm/membrane_core/Membrane.ParentSpec.html).

Reviewer 2: I think Membrane.ParentSpec is not always used while returning from the Pipeline callback - [callback_return_t()](https://hexdocs.pm/membrane_core/Membrane.Pipeline.html#t:callback_return_t/0) describes the return type, which consists of [Membrane.Pipeline.Action.t()](https://hexdocs.pm/membrane_core/Membrane.Pipeline.Action.html#t:t/0) which can be forward_t(), spec_t(), log_metadata_t() and some more - and Membrane.ParentSpec.t() is used probably only in spec_t().

It might be worth describing that there are other actions that can be taken while returning from the callback (and provide the second link I have put there within that description)

Reviewer 2: I suggest writing something like this:

The main purpose of the `handle_init` callback is to prepare our pipeline. Preparing means that we need to specify all its elements as children and set up links between those children to describe the order in which data will flow through the pipeline. 
Pipeline's callbacks are expected to return a status and an optional list of actions to be taken. The action can be of one of the following [types]( https://hexdocs.pm/membrane_core/Membrane.Pipeline.Action.html#t:t/0).
Since we want to spawn children processes and link them, we will use the [`spec_t()` action](https://hexdocs.pm/membrane_core/Membrane.Pipeline.Action.html#t:spec_t/0) which is described with the use of `Membrane.ParentSpec` structure.

> >If concept of callbacks and behaviours is new to you, you should probably take some time to read about OTP in Elixir (especially a part starring GenServer and Supervisor). You can find proper guide [here](https://elixir-lang.org/getting-started/mix-otp/agent.html)

Reviewer: ```suggestion
>If the concept of callbacks and behaviours is new to you, you should probably take some time to read about OTP in Elixir (especially the part starring GenServer and Supervisor). You can find the proper guide [here](https://elixir-lang.org/getting-started/mix-otp/agent.html)
```

#### Elements

The elements we'd like to use to play our mp3 will be:

 - `Membrane.File.Source` - the Source module form Membrane file plugin that will read a file.
 - `Membrane.MP3.MAD.Decoder` - an mp3 decoder based on [MAD](https://www.underbit.com/products/mad/)

>  - `Membrane.FFmpeg.SWResample.Converter` - a converter based on [FFmdeg SWResample](https://ffmpeg.org/doxygen/trunk/group__lswr.html#details). We'll be needing it to resample our row audio stream from 24 to 16 bits.

Reviewer: ```suggestion
 - `Membrane.FFmpeg.SWResample.Converter` - a converter based on [FFmdeg SWResample](https://ffmpeg.org/doxygen/trunk/group__lswr.html#details). We'll be needing it to resample our raw audio stream from 24 to 16 bits.
```

 - `Membrane.Portaudio.Sink` - A Sink module that will be playing our music with [Portaudio](http://www.portaudio.com)

> > As you can see all elements we're using are not a part of `membrane_core`, but can be found in separate libraries. You can find a list of packages provided by Membrane team [here](https://membraneframework.org/guide/v0.7/packages.html). You can also learn how to write your own Element.

Reviewer: ```suggestion
> As you can see all the elements we're using are not part of `membrane_core`, but can be found in separate plugins. You can find a list of packages provided by the Membrane Team [here](https://membraneframework.org/guide/v0.7/packages.html). There you can also learn how to write your own Element.
```

> The full children declaration for our player will look like that:

Reviewer: ```suggestion
The full children declaration for our player will look like this:
```

```elixir
children = %{
  file: %Membrane.File.Source{location: path_to_mp3},
  decoder: Membrane.MP3.MAD.Decoder,
  converter: %Membrane.FFmpeg.SWResample.Converter{
    output_caps: %Membrane.Caps.Audio.Raw{
	  format: :s16le,
	  sample_rate: 48000,
	  channels: 2
	}
  },
  portaudio: Membrane.PortAudio.Sink
}
```

> The keys in that keyword list are just a names we gave to elements. We're going to need them when linking.

Reviewer: ```suggestion
The keys in the `children` map (`file`, `decoder`, `converter`, `portaudio`) are just convenient names we gave our elements to refer to them later. We're going to need them for linking.
```

#### Linking elements

> Now we should link them in proper order. Each membrane element is one of three types: Source, Sink or Filter. The main difference is that Source provides only output pads, Sink only input and Filter both input and output. That means Source element can only start pipelines (it's not prepared to receive any data from other elements), Sink can only end pipeline (it will not send any data to succeeding elements), and Filters can be in the middle (it will receive, process and send data further). In our case a links declaration will look like that: 

Reviewer: ```suggestion
Now we should link them in the proper order. Each Membrane Element can be one of three types: Source, Sink or Filter. The main difference is that Source provides only output pads, Sink only input and Filter both input and output pads. That means only a Source element start pipelines (it's not prepared to receive any data from other elements), Sink can only end pipelines (it will not send any data to subsequent elements), and Filters can be in the middle (they receive, process and send data further). In our case the links declaration will look like this: 
```

```elixir
links = [
  link(:file)
  |> to(:decoder)
  |> to(:converter)
  |> to(:portaudio)
]
```

> File Source read bytes from our mp3, sends them to decoder. Decoder, after decoding, sends them to converter. Converter, after conversion sends them to sink. Portaudio sink receives them and plays music through Portaudio 🎶

Reviewer: ```suggestion
The file Source reads bytes from our mp3 file and sends them to decoder. Decoder, after decoding, sends them to converter. Converter, after conversion sends them to our portaudio sink, which receives them and plays music through Portaudio 🎶
```

#### Parent Spec

> Tha last but not least is to take elements and links together into a proper structure:

Reviewer: ```suggestion
Last but not least we need group our elements and links together into a proper structure:
```

> A structure here is `Membrane.ParentSpec` [docs](https://hexdocs.pm/membrane_core/Membrane.ParentSpec.html). You can also declare here other options if needed. In our pipeline `ParentSpec` will contains only children elements and links between them:

Reviewer: ```suggestion
The structure here is `Membrane.ParentSpec` [docs](https://hexdocs.pm/membrane_core/Membrane.ParentSpec.html). You can also declare other options here if needed. In our pipeline `ParentSpec` will contains only children elements and links between them:
```

```elixir
spec = %ParentSpec{children: children, links: links}
```

> At the and we need to return a proper tuple from `handle_init`. We can choose  from options described [here](https://hexdocs.pm/membrane_core/Membrane.Pipeline.html#t:callback_return_t/0-return-values). They are common for all callbacks, but as we're initialising pipeline we need to choose a one wich declares [actions to take with the pipeline](https://hexdocs.pm/membrane_core/Membrane.Pipeline.Action.html). Our action will be [`spec`](https://hexdocs.pm/membrane_core/Membrane.Pipeline.Action.html#t:spec_t/0):

Reviewer: ```suggestion
At the end of the callback we need to return a proper tuple from `handle_init`. We can choose  from options described [here](https://hexdocs.pm/membrane_core/Membrane.Pipeline.html#t:callback_return_t/0-return-values). They are common for all callbacks, but as we're initialising a pipeline we need to choose one wich declares [actions to take within the pipeline](https://hexdocs.pm/membrane_core/Membrane.Pipeline.Action.html). Our action will be [`spec`](https://hexdocs.pm/membrane_core/Membrane.Pipeline.Action.html#t:spec_t/0):
```

```elixir
{{:ok, spec: spec}, %{}}
```
We're passing empty map for state, as we don't need anything to be stored as state.

> After that a full modul will look like that:

Reviewer: ```suggestion
After that the full module will look like this:
```

```elixir
defmodule Hello do

  use Membrane.Pipeline

  @impl true
  def handle_init(path_to_mp3) do
	children = %{
	  file: %Membrane.File.Source{location: path_to_mp3},
	  decoder: Membrane.MP3.MAD.Decoder,
	  converter: %Membrane.FFmpeg.SWResample.Converter{
	    output_caps: %Membrane.Caps.Audio.Raw{
		  format: :s16le,
		  sample_rate: 48000,
		  channels: 2
		}
	  },
	  portaudio: Membrane.PortAudio.Sink
    }

	links = [
	  link(:file)
	  |> to(:decoder)
	  |> to(:converter)
	  |> to(:portaudio)
	]

	spec = %ParentSpec{children: children, links: links}

	{{:ok, spec: spec}, %{}}
  end
end

> ```

Reviewer 2: I am just wondering - Do we want to once again provide the full code snippet? Or perhaps it would be better not to put the whole snippet on the top of the tutorial as it will be put there?

Author: Yes, the first one is just a demonstration of how it will look, the last one is the result. I know it's repetition but it's worth in my opinion to repeat it, as it's first usage example of Membrane. We'll skip the demonstration in next tutorials.

### Running a pipeline

> Your pipeline you can start from any place in the code, but you can easily start it using Elixir interactive console:

Reviewer: ```suggestion
You can start your pipeline from any place in the code but it's convenient to use Elixir's interactive console:
```

Also note that while placing it anywhere works fine - such a call will only execute during compilation of the source file containing it.

```bash
$ iex -S mix
```

First, create a pipeline process:
```elixir
{:ok, pid}  =  Hello.start_link("/path/to/mp3")
```
Then, let it play:
```elixir
Hello.play(pid)
```

Our [demo with this tutorial](https://github.com/membraneframework/membrane_demo/tree/master/simple_pipeline) contains a file you can play, or you may want to use some [proper "Hello!" recording](https://upload.wikimedia.org/wikipedia/commons/transcoded/6/6f/Voyager_Golden_Record_greeting_in_English.ogg/Voyager_Golden_Record_greeting_in_English.ogg.mp3).

> The given mp3 file should be played on the default device in your system. Please use mp3 that has no ID3 or ID3v2 tags.

Reviewer: ```suggestion
The specified mp3 file should now play on the default device in your system. Please use mp3 that has no ID3 or ID3v2 tags.
```

end
```

The main purpose of the `handle_init` callback is to prepare our pipeline. Preparing means that we need to specify all its elements as children and set up links between those children to describe the order in which data will flow through the pipeline.

> Pipeline's callbacks are expected to return a status and an optional list of actions to be taken. The action can be of one of the following [types](https://hexdocs.pm/membrane_core/Membrane.Pipeline.Action.html#t:t/0).

Reviewer 2: ```suggestion
Pipeline's callbacks are expected to return status, an optional list of actions to be taken, and an updated state in which the pipeline will be after that callback. The action can be of one of the following [types](https://hexdocs.pm/membrane_core/Membrane.Pipeline.Action.html#t:t/0).
```

While suggesting previously, I have forgotten that the callback should also return an updated state - probably it is worth mentioning, just not to cause any misunderstandings.
